source/Halibut.Tests/TraceLogFileLogger.cs

```diff
@@ -1,25 +1,82 @@
 ﻿using System;
+using System.Collections.Generic;
 using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
 
 namespace Halibut.Tests
 {
     public class TraceLogFileLogger : IDisposable
     {
-        string tempFilePath = Path.GetTempFileName();
+        readonly AsyncQueue<string> queue = new AsyncQueue<string>();
+        readonly string tempFilePath = Path.GetTempFileName();
         string testHash;
 
+        readonly TaskCompletionSource<bool> noMoreDelaysToWritingLogs;
+        readonly Task noMoreDelaysTask;
+        readonly Task writeDataToDiskTask;
+
+        public TraceLogFileLogger()
+        {
+            noMoreDelaysToWritingLogs = new TaskCompletionSource<bool>();
+            noMoreDelaysTask = noMoreDelaysToWritingLogs.Task;
+            writeDataToDiskTask = WriteDataToFile();
+        }
+
         public void SetTestHash(string testHash)
         {
             this.testHash = testHash;
         }
         
         public void WriteLine(string logMessage)
         {
-            File.AppendAllLines(tempFilePath, new[] { logMessage });
+            queue.Enqueue(logMessage);
+        } 
+        
+        async Task WriteDataToFile()
+        {
+            while (true)
+            {
+                // Don't hammer the disk, let some log message queue up before writing them.
+                if (!noMoreDelaysTask.IsCompleted)
+                {
+                    await Task.WhenAny(Task.Delay(5), noMoreDelaysTask);
+                }
+
+                var list = new List<string>();
+                list.Add(await queue.DequeueAsync());
+                while (queue.TryDequeue(out var log))
+                {
+                    list.Add(log);
+                }
+
+                using (var fileAppender = new StreamWriter(tempFilePath, true, Encoding.UTF8, 8192))
+                {
+                    foreach (var logLine in list)
+                    {
+                        await fileAppender.WriteLineAsync(logLine);
+                    }
+
+                    await fileAppender.FlushAsync();
+                }
+            }
         }
+        
+        void AllowSomeTimeForTheQueueToEmpty()
+        {
+            noMoreDelaysToWritingLogs.SetResult(false);
+            for (int i = 0; i < 100; i++)
```

Comment:
Why do we only wait a little, instead of waiting for all log entries to be flushed?

Reply:
Is it because the same loop is used for all tests, so we don't actually ever shut it down?

Reply:
But if it's the same task for all tests, then we never reset `SetResult`. Maybe that's not the answer then

Reply:
Yeah good question, I was wondering what should we do if all the logs never get written because IO is too slow or something keeps spamming out to the logs.

Maybe waiting 100ms is not long enough, and instead we should wait longer? Or we could just wait forever and hope we can eventually have everything written....

Reply:
Each tests gets its own `TraceLogFileLogger`

Reply:
My gut feel is that we should terminate the writing loop thread/task, wait for that to finish, then write whatever is in the queue. And simply wait for all that to happen.

Reply:
The thing is, things can write to the logs after this method is called. Actually they do right after we call this method we log another thing :D

Reply:
OK updated it to wait for the write task to finish.